public/app/features/query/state/DashboardQueryRunner/SnapshotWorker.ts

```diff
@@ -0,0 +1,42 @@
+import { Observable, of } from 'rxjs';
+import { cloneDeep } from 'lodash';
+import { AnnotationEvent } from '@grafana/data';
+
+import { DashboardQueryRunnerOptions, DashboardQueryRunnerWorker, DashboardQueryRunnerWorkerResult } from './types';
+import { emptyResult } from './utils';
+
+export class SnapshotWorker implements DashboardQueryRunnerWorker {
+  canWork({ dashboard }: DashboardQueryRunnerOptions): boolean {
+    const snapshots = dashboard.annotations.list.find((a) => a.enable && Boolean(a.snapshotData));
+    return Boolean(snapshots);
+  }
+
+  work(options: DashboardQueryRunnerOptions): Observable<DashboardQueryRunnerWorkerResult> {
+    if (!this.canWork(options)) {
+      return emptyResult();
+    }
```

Comment:
Do we need this guard, since the rest of the function will effectively return `emptyResult` even if `dashAnnotations` and `snapshots` are empty?

Reply:
Not sure I follow you now, can you elaborate? This will return snapshotted annotations as far as I can see.

Reply:
So, if I understand correctly, `this.canWork` checks if `dashboard.annotations.list` has some value `a` such that `a.enable` is true, and that it has some snapshotData, and if this isn't the case, it returns `emptyResult()`, which is `of({ annotations: [], alertStates: [] })`.
But if we remove the guard
```ts
    if (!this.canWork(options)) {
      return emptyResult();
    }
```
then `getAnnotationsFromSnapshot` gets called regardless. Like in `canWork`, we check if `dashboard.annotations.list` has a value `a` such that `a.enable` is true, and that it has some snapshotData.
But even if it doesn't (i.e. `snapshots === []`), when it's reduced, `annotations` should just equal `[]`, and `getAnnotationsFromSnapshot` will return [], and so `work` will return `of({ annotations: [], alertStates: []})`, the same as `emptyResult()`.
It doesn't really matter, just removing the guard would mean there's only one path through the code which might be a little easier to read, but it's not a big deal at all.

Reply:
Hmm, maybe in the future the `canWork`will have other logic so I think I'll keep it for now